### ISSUE TITLE
fix(modeling): fake e adapter do LightGBM compartilham 49 linhas idênticas de validação

### ISSUE THREAD

**Issue by MarceloSanC** [MarceloSanC/financial-forecasting#71]
## Contexto

`src/financial_forecasting/features/modeling/adapters/out/lightgbm/lightgbm_quantile_trainer.py:275-323`
e `tests/fakes/features/modeling/in_memory_quantile_model_trainer.py:107-155`
são **49 linhas idênticas** — verificado por `diff`, saída vazia. É o maior bloco
duplicado do repositório, maior que o dos baselines (#66), e corresponde a ~32%
do fake.

O bloco é o `_validate_structure`: as validações estruturais da fronteira do port
(faixas de índices, alinhamento de labels por horizonte, consistência da grade de
quantis). Isso é **regra de contrato**, não detalhe de implementação da
biblioteca — e por estar duplicado, o contract test parametrizado `[fake, real]`
executa duas cópias da mesma validação e não pode detectar divergência entre elas.

Mesma patologia de #66 (baselines) e #70 (hasher). Este é o terceiro caso, e o
maior.

## O quê

Extrair a validação estrutural para o domínio de `modeling`, e fazer adapter e
fake delegarem — deixando no adapter apenas o que é específico do LightGBM
(construção de `Dataset`, parâmetros do booster, o `train`) e no fake apenas a
aritmética stdlib que substitui o treino.

## Como implementar

1. Criar (ou estender) um módulo de domínio com `validate_training_structure(...)`
   — função pura, stdlib-only, sobre primitivos/`collections.abc`.
2. Adapter e fake passam a chamá-la na entrada de `train_and_predict`.
3. Golden test antes de mexer, travando que a movimentação não altera
   comportamento observável (é refactor puro).
4. O contract test passa a exercitar o que de fato difere: o treino.

Se a #66 for feita antes, avaliar se as duas validações estruturais (baselines e
GBM) compartilham forma suficiente para viver no mesmo módulo — **sem forçar**:
unificar duas regras que só se parecem é pior que duplicá-las.

## Fundamentação

- **Fowler, ContractTest**: o teste deve verificar que o dublê responde como o
  real responderia. Duas cópias do mesmo código não sustentam essa proposição.
- **Fowler, TestDouble** (Meszaros): fake toma *"some shortcut"* — singular e
  localizado. O shortcut legítimo aqui é o treino, não a validação.
- **Evans, LAYERED ARCHITECTURE**: *"Concentrate all the code related to the
  domain model in one layer."* Validação de contrato de treino é regra, e regra
  não pertence a dois adapters.
- **Fowler, AnemicDomainModel**: *"The logic that should be in a domain object is
  domain logic — validations, calculations, business rules."*

## Critério de pronto

- [ ] Comparação estrutural adapter↔fake sem bloco de lógica compartilhada ≥15 linhas
- [ ] Golden test provando que nenhuma saída mudou
- [ ] Contract test exercitando o treino, não a validação
- [ ] `make check` verde

## Não-objetivos

- Mudar as validações em si — é movimentação, não mudança de comportamento.
- Unificar com a validação dos baselines sem evidência de que são a mesma regra.

## Momento previsto

**Junto com #66**, que é o mesmo trabalho no BC vizinho e provavelmente a mesma
sessão. Ambas são pré-requisito para que o gate `check_fake_parity` da #62 possa
ser ligado sem produzir uma lista de exceções que esvazia o gate.

## Referências

- Issue #66 (baselines — mesma patologia) · Issue #70 (hasher) · Issue #62 (o gate)
- `docs/LAYOUT.md` §7 — "A fake e a implementação real devem passar nos mesmos testes"
- https://martinfowler.com/bliki/ContractTest.html

**Comment by WhakaDev1** [comment]
`[finding]` — achado medido durante a execução (branch `refactor/66-emission-and-validation-to-domain`), registrado antes do fechamento.

**Existe um QUARTO caso da mesma patologia: o TFT trainer.**

Esta issue se declara "o terceiro caso, e o maior" (depois de #66 e #70). Rodando a mesma comparação estrutural no par que sobrou do BC `modeling`:

```
src/.../adapters/out/pytorch_forecasting/pf_tft_trainer.py
tests/fakes/features/modeling/in_memory_tft_trainer.py

EQUAL  17 linhas  A[251:268] <-> B[116:133]   <- assinatura do port (não é lógica)
EQUAL  16 linhas  A[147:163] <-> B[72:88]     <- _validate_contiguous INTEIRO
EQUAL  10 linhas  A[275:285] <-> B[156:166]   <- kwargs de delegação
EQUAL   9 linhas  A[407:416] <-> B[229:238]
TOTAL em blocos >= 5: 65 linhas
```

`_validate_contiguous` (faixa contígua e crescente de índices de decisão, C4) e `_eligible` estão duplicados literalmente entre adapter e dublê. É a mesma forma dos outros três: validação estrutural da fronteira do port morando nos dois lados, com o contract test `[fake, real]` rodando duas cópias.

Para efeito da **#62**: com #66 e #71 fechadas, o par do TFT é o que ainda produziria exceção na lista do `check_fake_parity`. O par do `HyperparameterSearch`/Optuna está limpo (maior bloco: 13 linhas, e é assinatura). Ou seja, ligar o gate depois desta PR exige exatamente **uma** exceção — a do TFT — ou fechar o TFT antes.

Não abri issue nova para não fragmentar sem decisão humana; o dado está aqui para quem for triar.

---

**Nota metodológica sobre o buraco medido**, que vale para as três issues da série: o gap não é só "duas cópias". A suíte parametrizada `[fake, real]` roda a mesma asserção separadamente em cada perna e **nunca compara uma com a outra**. Medido na base pré-refactor (`f93bc4f`), removendo `test_rows` do laço de largura **só no adapter**:

```
tests/contract/features/modeling/test_quantile_model_trainer_contract.py
  -> 30 passed  (VERDE)
```

Agravado pelo `match=r"C4"` dos testes de borda, que passa para qualquer mensagem C4 e não distingue um guard trocado por outro. Fechado nesta branch com `test_both_legs_reject_the_same_c4_input_with_the_same_message`, cobrindo as três posições do laço de largura e as duas do de rótulo (30 -> 40 testes).

**Comment by MarceloSanC** [comment]
`[correção]` — a "nota metodológica" do comentário acima ("a suíte `[fake, real]` nunca compara uma perna com a outra") foi refutada com controles na auditoria do PR #74: o verde era ausência de input de teste para o guard mutado, não incapacidade estrutural da suíte parametrizada. Detalhe e tabela em https://github.com/MarceloSanC/financial-forecasting/issues/62 (comentário de 2026-09-12) e no veredito em https://github.com/MarceloSanC/financial-forecasting/pull/74#issuecomment-5648467619. Os testes cruzados permanecem válidos (fecham inputs ausentes); só a justificativa muda.
